1

```diff
@@ -0,0 +1,87 @@
+import assert from 'assert';
+import axios from 'axios';
+import qs from 'qs';
+
+export const ENDPOINTS = {
+  REPORT_DATA: 'lib/data.php',
+  REPORT_CONFIG: 'lib/settings.php',
+  DATA_DICTIONARY: 'lib/data-dictionary.php'
+};
+
+/**
+ * Constructs a service that fetches data for the current project from REDCap.
+ *
+ * @param {Object} assetUrls - key/value pairs where the key is a relative file path
+ *   like `lib/data.php`, and the value is the corresponding external module URL
+ *   like `http://localhost/redcap/api/?type=module&prefix=consort-report&page=lib%2Fdata&pid=42`
+ * @return {Object} an object encapsulating REDCap HTTP requests
+ */
+export default function createDataService(assetUrls) {
+  assert(assetUrls, 'Asset URL object is required');
+  Object.keys(ENDPOINTS).forEach(key => {
+    const endpoint = ENDPOINTS[key];
+    assert(assetUrls[endpoint], `A URL for ${endpoint} is required`)
+  });
+
+  return {
+    reportDataUrl: assetUrls[ENDPOINTS.REPORT_DATA],
+    reportConfigUrl: assetUrls[ENDPOINTS.REPORT_CONFIG],
+    dataDictionaryUrl: assetUrls[ENDPOINTS.DATA_DICTIONARY],
+
+    /**
+     * Extracts data returned by the request.
+     * @param {Promise->Object} response - an axios response object.
+     * @return {Promise->Object} the response's data
+     * @private
+     */
+    _extractData(response) {
+      return response.data;
```

Comment:
Since all of your requests do `axios.post(...).then(this._extractData)`, why not make a `_makeRequest(url, data, options)` method?

Reply:
I went ahead and refactored this but realized there aren't any data-service unit tests. I think I can borrow something from Vizr. I'll work on these in another branch.